sycl/test-e2e/Matrix/joint_matrix_out_bounds_impl.hpp

```diff
@@ -0,0 +1,153 @@
+     cgh.parallel_for<class imatrix>(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [=](nd_item<2> spmd_item) [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<sub_group, bfloat16, use::a, TM, TK, layout::row_major>
+               sub_a;
+
+           // For B, since current implementation does not support non-packed
+           // layout, users need to specify the packed_b layout.
+           // By default, the layout is row_major
+           joint_matrix<sub_group, bfloat16, use::b, TK, TN,
+                        ext::intel::experimental::matrix::layout::packed>
+               sub_b;
+           joint_matrix<sub_group, float, use::accumulator, TM, TN> sub_c;
+           joint_matrix_fill(sg, sub_c, 1);
+           for (int k = 0; k < K; k += TK) {
+             joint_matrix_load(sg, sub_a, pA + (sg_startx * TM) * K + k, K);
+             // Assume we alreay in vnni format.
+             joint_matrix_load(
+                 sg, sub_b, pB + (k) * (N) + sg_starty / SG_SZ * TN * 2, N * 2);
+             sub_c = joint_matrix_mad(sg, sub_a, sub_b, sub_c);
+           }
+           joint_matrix_store(
+               sg, sub_c, pC + (sg_startx * TM) * N + sg_starty / SG_SZ * TN, N,
+               layout::row_major);
+         }); // parallel for
+   }).wait();
+}
+
+void matrix_multiply_ref(bfloat16 *A, bfloat16 *B, float *C, int MATRIX_M,
+                         int MATRIX_N, int MATRIX_K) {
+  for (unsigned int i = 0; i < MATRIX_M; i++) {
+    for (unsigned int k = 0; k < MATRIX_K; k++) {
+      for (unsigned int j = 0; j < MATRIX_N; j++) {
+        C[i * MATRIX_N + j] +=
+            make_fp32(A[i * MATRIX_K + k]) * make_fp32(B[k * MATRIX_N + j]);
+      }
+    }
+  }
+}
+
+template <typename T>
+void matrix_vnni(unsigned int rows, unsigned int cols, T *src, T *dest,
+                 unsigned int vnniFactor = 2) {
+  for (unsigned int i = 0; i < rows / vnniFactor; i++) {
+    for (unsigned int j = 0; j < cols; j++) {
+      for (unsigned int k = 0; k < vnniFactor; k++) {
+        dest[i * cols * vnniFactor + j * vnniFactor + k] =
+            src[(i * vnniFactor + k) * cols + j];
+      }
+    }
+  }
+}
+
+int main() {
+  // there will be 14*24 out of bounds in A matrix
+  static constexpr size_t MATRIX_M = 1024 + 14;
+  static constexpr size_t MATRIX_N = 1024;
+  static constexpr size_t MATRIX_K = 1024 + 24;
+  queue q;
```

Comment:
q is used only inside matrix_multiply kernel, so I suggest to get rid of unnecessary parameter passing